Add additional include and exclude options

matthayes opened the pull request:
## Proposed changes

This addresses #323 by adding these additional options:

* `--use-gitignore/--no-use-gitignore` for controlling whether `.gitignore` is used to provide exclude patterns
* `--force-include` to force a directory to be included, even if it would otherwise be ignored by other rules (such as those provided by a `.gitignore` file)
* `--force-include-pattern` to force files/directories to be included according to the pattern, even if they would otherwise be ignored by other rules (such as those provided by a `.gitignore` file)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Further comments

This does not change any existing behavior but merely provides additional options for controlling what to sync.

renardeinside commented on dbx/commands/sync.py:
```diff
@@ -94,48 +94,73 @@ def sync():
+    exclude_patterns.extend(subdirs_to_patterns(source, exclude_dirs))
+    force_include_patterns.extend(subdirs_to_patterns(source, force_include_dirs))
 
     gitignore_path = os.path.join(source, ".gitignore")
```
@matthayes could you please use [pathlib](https://docs.python.org/3/library/pathlib.html) instead of `os.path` and `with open(..)`?

matthayes replied:
done

renardeinside commented on dbx/commands/sync.py:
```diff
@@ -94,48 +94,73 @@ def sync():
-    gitignore_path = os.path.join(source, ".gitignore")
-    if os.path.exists(gitignore_path):
+    gitignore_path = Path(source) / ".gitignore"
+    if use_gitignore and os.path.exists(gitignore_path):
```
instead of: `os.path.exists(gitignore_path)` please use `gitignore_path.exists()`

matthayes replied:
done

renardeinside commented on dbx/commands/sync.py:
```diff
@@ -94,48 +94,73 @@ def sync():
+    gitignore_path = Path(source) / ".gitignore"
+    if use_gitignore and os.path.exists(gitignore_path):
         dbx_echo(f"Ignoring patterns from {gitignore_path}")
         with open(gitignore_path, encoding="utf-8") as f:
```
instead of `open()` please use:
`lines = gitignore_path.read_text(encoding="utf-8").splitlines()`

matthayes replied:
done

renardeinside reviewed (approved):
LGTM!

renardeinside commented:
@matthayes , 
everything seems good to me, please reflect the changes in CHANGELOG and feel free to merge

matthayes commented:
Thanks @renardeinside !  I've updated the changelog.  By the way, I discovered a bug, which I've addressed in the new commit below:

https://github.com/databrickslabs/dbx/pull/351/commits/03eff6d47fbecb3d93be93e8442ba6787fb84224

Also it looks like I don't have control to merge.
